Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,261 @@
+  {
+    this(divergence, selfDivergence, DEFAULT_DAMPING_FACTOR);
+  }
+
+  /**
+   * Creates a new instance of AffinityPropagation.
+   *
+   * @param  divergence The divergence function to use to determine the
+   *         divergence between two examples.
+   * @param  selfDivergence The maxValue for self-divergence to use, which
+   *         controls the number of clusters created.
+   * @param  dampingFactor The damping factor (lambda). Must be between 0.0
+   *         and 1.0.
+   */
+  private OptimizedAffinityPropagation(
+      DivergenceFunction<? super T, ? super T> divergence,
+      double selfDivergence,
+      double dampingFactor)
+  {
+    this(divergence, selfDivergence, dampingFactor, DEFAULT_MAX_ITERATIONS);
+  }
+
+  /**
+   * Creates a new instance of AffinityPropagation.
+   *
+   * @param  divergence The divergence function to use to determine the
+   *         divergence between two examples.
+   * @param  selfDivergence The maxValue for self-divergence to use, which
+   *         controls the number of clusters created.
+   * @param  dampingFactor The damping factor (lambda). Must be between 0.0
+   *         and 1.0.
+   * @param  maxIterations The maximum number of iterations.
+   */
+  private OptimizedAffinityPropagation(
+      DivergenceFunction<? super T, ? super T> divergence,
+      double selfDivergence,
+      double dampingFactor,
+      int maxIterations)
+  {
+    super(divergence, selfDivergence, dampingFactor, maxIterations);
+  }
+
+  @Override
+  protected boolean initializeAlgorithm() {
+    if (this.getData() == null || this.getData().isEmpty())
+    {
+      // Make sure that the data is valid.
+      return false;
+    }
+
+    // Initialize the main data for the algorithm.
+    setExamples(new ArrayList<>(getData()));
+    setSimilarities(new double[exampleCount][exampleCount]);
+    setResponsibilities(new double[exampleCount][exampleCount]);
+    setAvailabilities(new double[exampleCount][exampleCount]);
+
+    double[] divergenciesArray = new double[exampleCount * (exampleCount+1)/2];
+    boolean computeSelfDivergence = Double.isNaN(getSelfDivergence());
+
+    // Compute the similarity matrix.
+    fillSimilarityStructures(divergenciesArray);
+
+    if(computeSelfDivergence) {
+      double median = computeMedian(divergenciesArray);
+      setSelfDivergence(median);
+    }
+
+    fillSimilarityDiagonal();
+
+    // Initialize the assignments to -1, the changed count, and the
+    // clusters.
+    this.setAssignments(new int[exampleCount]);
+    this.setChangedCount(exampleCount);
+    this.setClusters(new HashMap<>());
+    for (int i = 0; i < exampleCount; i++)
+    {
+      this.assignments[i] = -1;
+    }
+
+    // Ready to learn.
+    return true;  }
+
+  private void fillSimilarityDiagonal() {
+    // Set the self similarity based on the self divergence.
+    for (int i = 0; i < exampleCount; i++)
+    {
+      similarities[i][i] = -getSelfDivergence();
+    }
+  }
+

```

Comment:
I think the divergenciesArray (should be divergencesArray?) is a complex enough (but cool!) trick that it's worth j function Javadoc explaining its internal referencing specifically.  It appears to be a 1-d vector of an upper-triangular 2-d array, right?

Reply:
I didn't get this point. I simply extended the self divergence hanling, set to the median of all the divergencies, when it's not specified by the client. What do i have to explain in a more specific way? Thank you.